# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsIFrame.h
# Commit: c428c3ad7725
# Full Hash: c428c3ad7725593f33b4ea2da9851be31be0c1df
# Author: Martin Robinson <mrobinson@igalia.com>
# Date: 2022-11-28 15:41:50
# Regressor Bug: 1791759
# File Overlap Count: 1
# Description:
#   Bug 1791759 - Add support for `content-visibility: auto` r=emilio
#   
#   This change adds support for `content-visibilty: auto` as well as
#   showing and hiding content based on the relevancy of the content as
#   defined in the specification. Changes to relevancy are handled by
# ==============================================================================

diff -r ce553434720c -r c428c3ad7725 layout/generic/nsIFrame.h
--- a/layout/generic/nsIFrame.h	Mon Nov 28 09:45:39 2022 +0000
+++ b/layout/generic/nsIFrame.h	Mon Nov 28 09:53:06 2022 +0000
@@ -536,6 +536,7 @@
   using ReflowOutput = mozilla::ReflowOutput;
   using Visibility = mozilla::Visibility;
   using LengthPercentage = mozilla::LengthPercentage;
+  using ContentRelevancy = mozilla::ContentRelevancy;
 
   using nsDisplayItem = mozilla::nsDisplayItem;
   using nsDisplayList = mozilla::nsDisplayList;
@@ -3182,10 +3183,33 @@
   bool IsContentDisabled() const;
 
   /**
+   * Whether the content-visibility CSS property applies to this frame.
+   */
+  bool IsContentVisibilityPropertyApplicable() const;
+
+  enum class IncludeContentVisibility {
+    Auto,
+    Hidden,
+  };
+
+  constexpr static mozilla::EnumSet<IncludeContentVisibility>
+  IncludeAllContentVisibility() {
+    return {IncludeContentVisibility::Auto, IncludeContentVisibility::Hidden};
+  }
+
+  /**
+   * Returns true if this frame's `content-visibility: auto` element is
+   * considered relevant content.
+   */
+  bool IsContentRelevant() const;
+
+  /**
    * Whether this frame hides its contents via the `content-visibility`
    * property.
-   */
-  bool HidesContent() const;
+   * @param aInclude specifies what kind of `content-visibility` to include.
+   */
+  bool HidesContent(const mozilla::EnumSet<IncludeContentVisibility>& =
+                        IncludeAllContentVisibility()) const;
 
   /**
    * Whether this frame hides its contents via the `content-visibility`
@@ -3198,8 +3222,11 @@
   /**
    * Returns true if this frame is entirely hidden due the `content-visibility`
    * property on an ancestor.
-   */
-  bool IsHiddenByContentVisibilityOnAnyAncestor() const;
+   * @param aInclude specifies what kind of `content-visibility` to include.
+   */
+  bool IsHiddenByContentVisibilityOnAnyAncestor(
+      const mozilla::EnumSet<IncludeContentVisibility>& =
+          IncludeAllContentVisibility()) const;
 
   /**
    * Returns true is this frame is hidden by its first unskipped in flow
@@ -3208,6 +3235,27 @@
   bool IsHiddenByContentVisibilityOfInFlowParentForLayout() const;
 
   /**
+   * Whether or not this frame's content is a descendant of a top layer element
+   * used to determine if this frame is relevant content for
+   * `content-visibility: auto`.
+   */
+  bool IsDescendantOfTopLayerElement() const;
+
+  /**
+   * Returns true if this frame has a SelectionType::eNormal type selection in
+   * somewhere in its subtree of frames. This is used to determine content
+   * relevancy for `content-visibility: auto`.
+   */
+  bool HasSelectionInSubtree();
+
+  /**
+   * Update the whether or not this frame is considered relevant content for the
+   * purposes of `content-visibility: auto` according to the rules specified in
+   * https://drafts.csswg.org/css-contain-2/#relevant-to-the-user.
+   */
+  void UpdateIsRelevantContent(const ContentRelevancy& aRelevancyToUpdate);
+
+  /**
    * Get the "type" of the frame.
    *
    * @see mozilla::LayoutFrameType